File: client-samples/ios-visionos/StreamKit/Sources/StreamKit/Backends/LiveKit/ByteStreamTransport.swift
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import Foundation
import LiveKit

internal struct ByteStreamWireOptions: Sendable {
    let topic: String
    let attributes: [String: String]
    let destinationIdentities: [Participant.Identity]
    let mimeType: String?
    let name: String?
    let totalSize: Int?
}

internal struct ByteStreamConnection: @unchecked Sendable {
    let room: Room
    let generation: UInt64
}

internal enum ByteStreamTransportError: Error {
    case connectionChanged
}

private actor ByteStreamCloseRace {
    private var completed = false
    private var waiter: CheckedContinuation<Void, Never>?

    func wait() async {
        if completed { return }
        await withCheckedContinuation { waiter = $0 }
    }

    func finish() {
        guard !completed else { return }
        completed = true
        waiter?.resume()
        waiter = nil
    }
}

internal enum LiveKitByteStreamWriter {
    static func sendBytes(
        _ data: Data,
        options: ByteStreamWireOptions,
        connection: ByteStreamConnection,
        isConnectionActive: (ByteStreamConnection) -> Bool
    ) async throws -> String {
        try await send(
            source: .data(data),
            options: options,
            connection: connection,
            isConnectionActive: isConnectionActive
        )
    }

    static func sendFile(
        _ fileURL: URL,
        options: ByteStreamWireOptions,
        connection: ByteStreamConnection,
        isConnectionActive: (ByteStreamConnection) -> Bool
    ) async throws -> String {
        try await send(
            source: .file(fileURL),
            options: options,
            connection: connection,
            isConnectionActive: isConnectionActive
        )
    }

    private enum Source {
        case data(Data)
        case file(URL)
    }

    private static func send(
        source: Source,
        options: ByteStreamWireOptions,
        connection: ByteStreamConnection,
        isConnectionActive: (ByteStreamConnection) -> Bool
    ) async throws -> String {
        guard isConnectionActive(connection) else {
            throw ByteStreamTransportError.connectionChanged
        }
        let writer = try await connection.room.localParticipant.streamBytes(
            options: StreamByteOptions(
                topic: options.topic,
                attributes: options.attributes,
                destinationIdentities: options.destinationIdentities,
                mimeType: options.mimeType,
                name: options.name,
                totalSize: options.totalSize
            )
        )
        do {
            switch source {
            case .data(let data):
                try await writer.write(data)
            case .file(let fileURL):
                let handle = try FileHandle(forReadingFrom: fileURL)
                defer { try? handle.close() }
                while let chunk = try handle.read(upToCount: 64 * 1024), !chunk.isEmpty {
                    try Task.checkCancellation()
                    guard isConnectionActive(connection) else {
                        throw ByteStreamTransportError.connectionChanged
                    }
                    try await writer.write(chunk)
                }
            }
            try Task.checkCancellation()
            try await writer.close()
            guard await writer.isOpen == false, isConnectionActive(connection) else {
                throw ByteStreamTransportError.connectionChanged
            }
        } catch {
            await closeAfterFailure(writer)
            throw error
        }
        return writer.info.id
    }

    private static func closeAfterFailure(_ writer: ByteStreamWriter) async {
        let race = ByteStreamCloseRace()
        let closeTask = Task {
            _ = try? await writer.close(reason: "StreamKit send failed")
            await race.finish()
        }
        let timeoutTask = Task {
            _ = try? await Task.sleep(nanoseconds: 2_000_000_000)
            await race.finish()
        }
        await race.wait()
        closeTask.cancel()
        timeoutTask.cancel()
    }
}

File: client-samples/android/app/src/main/java/com/nvidia/xrai/streamkitsample/streamkit/backends/livekit/ByteStreamTransport.kt
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.nvidia.xrai.streamkitsample.streamkit.backends.livekit

import io.livekit.android.room.Room
import io.livekit.android.room.datastream.StreamBytesOptions
import io.livekit.android.room.datastream.outgoing.ByteStreamSender
import io.livekit.android.room.datastream.outgoing.writeFile
import io.livekit.android.room.participant.Participant
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import java.io.File

internal data class ByteStreamWireOptions(
    val topic: String,
    val attributes: Map<String, String> = emptyMap(),
    val destinationIdentities: List<Participant.Identity> = emptyList(),
    val mimeType: String,
    val name: String,
    val totalSize: Long,
)

internal data class ByteStreamConnection(
    val room: Room,
    val generation: Long,
)

internal class ByteStreamConnectionChanged : IllegalStateException(
    "The byte stream is no longer on the active connection",
)

internal class LiveKitByteStreamWriter(
    private val snapshotConnection: () -> ByteStreamConnection?,
    private val isConnectionActive: (ByteStreamConnection) -> Boolean,
) {
    suspend fun sendBytes(data: ByteArray, options: ByteStreamWireOptions): String =
        send(options) { sender -> sender.write(data).getOrThrow() }

    suspend fun sendFile(file: File, options: ByteStreamWireOptions): String =
        send(options) { sender -> sender.writeFile(file).getOrThrow() }

    private suspend fun send(
        options: ByteStreamWireOptions,
        write: suspend (ByteStreamSender) -> Unit,
    ): String {
        val connection = snapshotConnection() ?: throw ByteStreamConnectionChanged()
        if (!isConnectionActive(connection)) throw ByteStreamConnectionChanged()
        val sender = connection.room.localParticipant.streamBytes(
            StreamBytesOptions(
                topic = options.topic,
                attributes = options.attributes,
                destinationIdentities = options.destinationIdentities,
                mimeType = options.mimeType,
                name = options.name,
                totalSize = options.totalSize,
            ),
        )
        try {
            write(sender)
            sender.close()
            if (sender.isOpen || !isConnectionActive(connection)) {
                throw ByteStreamConnectionChanged()
            }
        } catch (error: Throwable) {
            closeAfterFailure(sender)
            throw error
        }
        return sender.info.id
    }

    private suspend fun closeAfterFailure(sender: ByteStreamSender) {
        withContext(NonCancellable) {
            withTimeoutOrNull(2_000) {
                runCatching { sender.close("StreamKit send failed") }
            }
        }
    }
}
